c10/core/Device.h

```diff
@@ -32,8 +33,10 @@ struct C10_API Device final {
 
   /// Constructs a new `Device` from a `DeviceType` and an optional device
   /// index.
-  /* implicit */ Device(DeviceType type, DeviceIndex index = -1)
-      : type_(type), index_(index) {
+  /* implicit */ Device(DeviceType type, int64_t index = -1) : type_(type) {
+    TORCH_CHECK(std::numeric_limits<DeviceIndex>::min() <= index, "Device index too small!");
+    TORCH_CHECK(index <= std::numeric_limits<DeviceIndex>::max(), "Device index too large!");
```

Comment:
This path is hot, so we're not going to take this extra error checking. See https://github.com/pytorch/pytorch/pull/47894